### DISCUSSION TITLE
Prevent overlap between multiple tasks

### DISCUSSION THREAD

**Discussion by DuelingCats** [jamesmh/coravel#350] [Q&A]
From reading the docs on [PreventOverlapping](https://docs.coravel.net/Scheduler/#prevent-overlapping-tasks), it sounds like the functionality is more targeted towards preventing a specific task from running concurrently. In my situation, I want to prevent two tasks from running concurrently in regards to itself and the other tasks, if that makes sense. So at any one time there should be only 1 task running 1 instance of said task.

Given the scenario, if I share the same key for `PreventOverlapping`, would this work or is the scheduler not designed to work in this way?

For example:
```
host.Services.UseScheduler(scheduler => 
{
    scheduler.Schedule<FirstInvocableTask>().EverySeconds(3).PreventOverlapping("SharedOverlapKey");
    scheduler.Schedule<SecondInvocableTask>().EverySeconds(30).PreventOverlapping("SharedOverlapKey");
});
```

**Comment by jamesmh** [comment]
Yes, the intent is to prevent overlapping on 1 invocable.

What would end up happening depends on when and how long each task runs. In the scenario you laid out the following would occur:

1. `FirstInvocableTask` will run at seconds 30 (within each minute). 
2. It will lock and prevent `SecondInvocableTask` from running at this point. 
3. This will occur every 30 seconds because both tasks/invocables are due at seconds 30.
4. `SecondInvocableTask` will therefore never run.

You'd have to come up with your own way to lock both tasks - either a semaphore, `lock`, or a distributed lock of some kind.

**Reply by DuelingCats** [reply]
I did some testing earlier today, and it appears to be working without a semaphore or lock. With the below example code, I don't see where the `LongRunningTask2` gets prevented from running with a shared key.

```
Task 1 complete at 10/10/2023 11:47:02 PM
Task 1 complete at 10/10/2023 11:47:04 PM
Task 1 complete at 10/10/2023 11:47:06 PM
Task 1 complete at 10/10/2023 11:47:08 PM
Task 2 complete at 10/10/2023 11:47:10 PM
Task 1 complete at 10/10/2023 11:47:10 PM
Task 1 complete at 10/10/2023 11:47:12 PM
Task 1 complete at 10/10/2023 11:47:14 PM
Task 1 complete at 10/10/2023 11:47:16 PM
Task 1 complete at 10/10/2023 11:47:18 PM
Task 2 complete at 10/10/2023 11:47:20 PM
Task 1 complete at 10/10/2023 11:47:20 PM
Task 1 complete at 10/10/2023 11:47:22 PM
Task 1 complete at 10/10/2023 11:47:24 PM
Task 1 complete at 10/10/2023 11:47:26 PM
Task 1 complete at 10/10/2023 11:47:28 PM
Task 2 complete at 10/10/2023 11:47:30 PM
Task 1 complete at 10/10/2023 11:47:30 PM
Task 1 complete at 10/10/2023 11:47:32 PM
Task 1 complete at 10/10/2023 11:47:34 PM
Task 1 complete at 10/10/2023 11:47:36 PM
Task 1 complete at 10/10/2023 11:47:38 PM
Task 2 complete at 10/10/2023 11:47:40 PM
Task 1 complete at 10/10/2023 11:47:40 PM
Task 1 complete at 10/10/2023 11:47:42 PM
Task 1 complete at 10/10/2023 11:47:44 PM
```

```
using Coravel;
using Coravel.Invocable;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.DependencyInjection;

var builder = Host.CreateApplicationBuilder(args);
builder.Services.AddScheduler();
builder.Services.AddTransient<LongRunningTask1>();
builder.Services.AddTransient<LongRunningTask2>();

var host = builder.Build();

host.Services.UseScheduler(scheduler =>
{
    scheduler.Schedule<LongRunningTask1>().EverySeconds(2).PreventOverlapping("SharedOverlapKey");
    scheduler.Schedule<LongRunningTask2>().EverySeconds(10).PreventOverlapping("SharedOverlapKey");
});
    
    
await host.RunAsync();

public class LongRunningTask1 : IInvocable
{
    public async Task Invoke()
    {

        await Task.Delay(50);
        Console.WriteLine($"Task 1 complete at {DateTime.Now}");
    }
}

public class LongRunningTask2 : IInvocable
{
    public async Task Invoke()
    {
        await Task.Delay(50);
        Console.WriteLine($"Task 2 complete at {DateTime.Now}");
    }
}
```

**Reply by jamesmh** [reply]
Looks like it's allowing the two tasks to go through. I'd have to check the code to see what's going on. But ya, this isn't the intended way to use `PreventOverlapping`.
